### PR TITLE
KM230 👌 Convert apply application to reconciliation pattern

### PR DESCRIPTION
**Pull request by deifyed**
## Description

Makes `apply application` use the reconciler/operator pattern as in `apply cluster`

## Motivation and Context

[KM230](https://trello.com/c/xbYLxDEe)

Julius likes this pattern. (Also it provides a more "natural" way to introduce more features to an application, f.eks ECR)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I have updated the release notes (for the next release).

### CODE REVIEW

File: cmd/okctl/apply_application.go [hunk 1]
```diff
@@ -91,15 +80,49 @@ func buildApplyApplicationCommand(o *okctl.Okctl) *cobra.Command {
...
-			err = services.ApplicationService.ScaffoldApplication(o.Ctx, scaffoldOpts)
+			reconciliationManager := reconciler.NewCompositeReconciler(spin,
+				reconciler.NewApplicationReconciler(services.ApplicationService),
+			)
+
+			reconciliationManager.SetCommonMetadata(&resourcetree.CommonMetadata{
+				Ctx:       o.Ctx,
+				Out:       o.Out,
+				ClusterID: *scaffoldOpts.ID,
+				// We should pass inn cluster fetched from state here when state rewrite is done
```
Comment (ivaruf) [inline cmd/okctl/apply_application.go:94]: Add TODO? Actually valid case :)

**Review by ivaruf — APPROVED** [review]
Looks great :+1:

File: pkg/controller/reconciler/application_reconciler.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+			HostedZoneID:     state.PrimaryHostedZoneID,
+			HostedZoneDomain: state.PrimaryHostedZoneDomain,
+			IACRepoURL:       a.commonMetadata.Declaration.Github.Repository,
+			Application:      state.Declaration,
+		})
+		if err != nil {
+			return ReconcilationResult{}, err
+		}
+	case resourcetree.ResourceNodeStateAbsent:
+		return ReconcilationResult{}, errors.New("deletion of applications is not implemented")
```
Comment (yngvark) [inline pkg/controller/reconciler/application_reconciler.go:55]: Should we add this as a Trello-task?

File: pkg/controller/reconciler/application_reconciler.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
+package reconciler
```
Comment (yngvark) [inline pkg/controller/reconciler/application_reconciler.go:1]: Have you considered putting this reconciler in a separate directory? E.g. `pkg/controller/reconciler/application/`, next to `pkg/controller/reconciler/cluster/` for the others?

**Review by yngvark — APPROVED** [review]
Looks good :+1:

File: pkg/controller/reconciler/application_reconciler.go (hunk 3 above)
Comment (deifyed) [reply pkg/controller/reconciler/application_reconciler.go:1]: Yeah, but couldn't really make up my mind. The certificate reconciler is for example used by both the cluster and applications. Meaning we'd need a `common`, `cluster` and an `application`.

Do chime in if you have any thoughts

File: pkg/controller/reconciler/application_reconciler.go (hunk 2 above)
Comment (deifyed) [reply pkg/controller/reconciler/application_reconciler.go:55]: Good idea

File: pkg/controller/reconciler/application_reconciler.go (hunk 3 above)
Comment (yngvark) [reply pkg/controller/reconciler/application_reconciler.go:1]: Right, I knew there was going to be some issues.

Slightly related: In convert.go we have both `CreateResourceDependencyTree` and `CreateApplicationResourceDependencyTree`, so we don't have any clear separation there either. So all together I think we can leave it as it is, it's simple enough.

Comment (deifyed) [reply pkg/controller/reconciler/application_reconciler.go:1]: Also a good point. 

I kinda settled on waiting until we got one more case for creating dependency trees / reconcilers to see if there was an obvious pattern

**Pull request merged**